README
NAME
    Castella - Sinatra like web application framework for PSGI/Plack

SYNOPSIS
      package  MyApp;
      use Castella;

      get '/' => run {
            my $self = shift;
            $self->stash->{message} = 'Hello Castella world!';
      };
      post '/' => run {
            my $self = shift;
            $self->stash->{message} = 'Castella is sweet!';
      };
      1;
      __DATA__
      @@ /
      <!DOCTYPE HTML>
      <html lang="ja">
      <head>
        <meta charset="UTF-8">
        <title>hogehoge</title>
      </head>
      <body>
        <: $message :>
        <form action="/" method="post">
            <button  type="submit">press me!</button>
        </form>
      </body>
      </html>

DESCRIPTION
    Castella is oreore waf which like Sinatra on Ruby

    Write code easily!

METHODS
AUTHOR
    Nishibayashi Takuji <takuji {at} senchan.jp>

SEE ALSO
LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Castella

0.01    Wed Apr 20 07:31:14 2011
        - original version


t/app/lib/TestApp.pm
package TestApp;
use Castella;

get '/' => run {
    my $self = shift;
    $self->stash->{message} = 'Hello Castella world!';
};

post '/' => run {
    my $self = shift;
    $self->stash->{message} = 'Hello Castella HogeHoge!';
};

1;
__DATA__
@@ /
<!DOCTYPE HTML>
<html lang="ja">
<head>
  <meta charset="UTF-8">
  <title>index</title>
</head>
<body>
    <p>
        <: $message :>
    </p>
    <form action="/" method="POST">
        <button type="submit">POST</button>
    </form>
</body>
</html>


lib/Castella/Lite.pm
package  Castella::Lite;
use Castella -base;

use parent qw/Class::Data::Inheritable/;

use Class::Accessor::Lite (
    new => 1,
    ro  => [qw/ req res stash /],
);
use Data::Section::Simple;
use Text::Xslate;

use Castella::Exception;
use Castella::Request;
use Castella::Response;

sub import {
    my $class  = shift;
    my $caller = caller;

    strict->import;
    warnings->import;

    {
        no strict 'refs';
        unshift @{"$caller\::ISA"}, $class;
    }
    #make action_table
    $caller->mk_classdata(action_table => []);
    #make attributes
    $caller->mk_classdata(attr => {});
    #Export DSL
    my @functions = qw/
        run
        action
        post
        get
    /;
    for my $function ( @functions ) {
        $class->export_method($caller, $function);
    }
}
sub run (&) { $_[0] }

sub add_action {
    my ($class, $regex, $options, $code, $method) = @_;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code , options => $options, method => $method};
}
sub action ($$) {
    my ($regex, $code) = @_;
    my $class = caller;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code , template => $regex};
}

sub post ($$) {
    my ($regex, $code) = @_;
    my $class = caller;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code, template => $regex, method => 'POST'};
}

sub get ($$) {
    my ($regex, $code) = @_;
    my $class = caller;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code, template => $regex, method => 'GET'};
}

sub get_router {
    return sub {
        my ($class, $env) = @_;
        my $path = $env->{PATH_INFO};
        my $code;
        for my $action ( @{$class->action_table} ) {
            if ( $path =~ $action->{regex} && ( !defined $action->{method} || $action->{method} eq $env->{REQUEST_METHOD} ) ) {
                $code = $action->{code};
                last;
            }

        }
        Castella::Exception::HTTP::NotFound->throw unless $code;
        my $c = $class->new(
            req   => Castella::Request->new($env),
            res   => Castella::Response->new,
            stash => {},
            env   => $env,
        );
        $code->($c);
        my $templates = Data::Section::Simple->new($class)->get_data_section;
        my $xslate = Text::Xslate->new(
            path => $templates,
        );
        my $body = $xslate->render($path,$c->stash);
        Castella::Exception::HTTP::Success->throw(
            res => [
                200,
                ['Content-Type' => 'text/html'],
                [$body]
            ],
        );
    };
}

1;
__END__
=encoding utf8

=head1 NAME

Castella::Lite - Sinatra like implementation for Castella

=head1 SYNOPSIS

  package  MyApp;
  use Castella::Lite -base;

=head1 DESCRIPTION

Castella::Lite is Sinatra like implementation for Castella

Write code more easily!

=head1 AUTHOR

Nishibayashi Takuji E<lt>takuji {at} senchan.jpE<gt>

=head1 SEE ALSO

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Castella/Utils.pm
package  Castella::Utils;
use strict;
use warnings;

use Exporter::Lite;

our @EXPORT = qw(
    export_method
    export_coderef
);

sub export_method {
    my ( $class, $target, $method_name ) = @_;
    if ( ref($target) ) {
        $target = ref($target);
    }
    my $code = $class->can($method_name);
    unless ($code) {
        Carp::croak("Method $class\::$method_name does not exists!");
    }
    export_coderef( $target, $method_name, $code );
}

sub export_coderef {
    my ( $target, $method_name, $code ) = @_;
    if ( ref($target) ) {
        $target = ref($target);
    }
    unless ( ref($code) eq 'CODE' ) {
        Carp::croak("This is not code reference! $code");
    }
    {
        no strict 'refs';    ## no critic
        *{"$target\::$method_name"} = $code;
    }
}

1;
__END__
=encoding utf8

=head1 NAME

Castella::Utils - Utilities for Castella

=head1 SYNOPSIS

  use Castella::Utils;

=head1 METHODS

=head2 export_method

Export method in current class.

  package  MyApp::Hoge;
  use MyApp -util;
  sub import {
      my $class  = shift;
      my $caller = caller;
      $class->export_coderef($caller,'foo');
  }
  sub foo {
    #Do something!
  }

  package  MyApp::Fuga;
  use MyApp::Hoge;

  sub hoge {
    foo();
  }

=head2 export_coderef

Export code reference.

  package  MyApp::Hoge;
  use MyApp -util;
  sub import {
      my $class  = shift;
      my $caller = caller;
      $class->export_metod($caller,'byaa',sub { 'umai' });
  }

  package  MyApp::Fuga;
  use MyApp::Hoge;

  sub hoge {
    byaa();
  }

=head1 AUTHOR

Nishibayashi Takuji E<lt>takuji {at} senchan.jpE<gt>

=head1 SEE ALSO

=cut


lib/Castella/Exception.pm
package Castella::Exception;
use strict;
use warnings;
use overload '""' => \&inspect;

use Carp ();
use Class::Accessor::Lite (
    new => 1,
    rw  => [qw/ msg /],
);
use Class::Load qw/:all/;

use Castella::Utils;

sub import {
    my $class  = shift;
    my $caller = caller;
    load_class('Castella::Exception::HTTP');
}

sub inspect {
    my $self = shift;

    sprintf("%s < msg: %s > at %s line %s.", ref($self), $self->msg || 'undefined' , $self->{caller}, $self->{line_number})

}

sub throw {
    my ($class, %args) = @_;

    if ( $class eq __PACKAGE__ ) {
        Carp::croak("$class can't use directly!");
    }

    my ($caller, $call_path, $line_number) = caller(1);
    $args{caller} = $call_path;
    $args{line_number} = $line_number;

    my $e = $class->new(%args);
    Carp::croak($e);
}

1;


lib/Castella/Exception/HTTP.pm
package  Castella::Exception::HTTP;
use strict;
use warnings;
use Castella::Utils;

use parent qw/Castella::Exception/;

use Class::Accessor::Lite (
    new => 1,
    rw => [qw/ res /],
);

sub response { croak("Please override response method!") }

package  Castella::Exception::HTTP::Success;
use parent qw/Castella::Exception::HTTP/;

sub response {
    my $self = shift;
    my $res  = $self->res;

    if( ref($res) eq 'ARRAY' ) {
        return $res;
    } elsif ( ref($res) eq 'HASH' ) {
        return [
            $res->{status},
            $res->{headers},
            [$res->{body}],
        ];
    } elsif ( ref($res) && $res->isa('Plack::Response') ) {
        return $res->finalize;
    } else {
        return [
            200,
            ['Content-Type' => 'text/html'],
            [$res],
        ];
    }
}



package  Castella::Exception::HTTP::NotFound;
use parent qw/Castella::Exception::HTTP/;

sub response {
    my $self = shift;
    my $res  = $self->res;
    my $headers;
    my $body;
    if (ref($res) eq 'HASH') {
        $headers = $res->{headers};
        $body    = $res->{body};
    } else {
        $body = $res;
    }
    $headers ||= [ 'Content-Type' => 'text/plain' ];
    $body    ||= '404 Not Found';

    return [
        404,
        $headers,
        [$body],
    ];
}


1;


lib/Castella.pm
package Castella;
use strict;
use warnings;
our $VERSION = '0.01';

use parent 'Class::Data::Inheritable';

use Class::Accessor::Lite (
    new => 1,
    ro  => [qw(req res stash)],
);

use Carp ();
use Class::Load ':all';
use Data::Section::Simple;
use Text::Xslate;
use Try::Tiny;
use UNIVERSAL::can;

use Castella::Exception;
use Castella::Request;
use Castella::Response;
use Castella::Utils;

sub import {
    my $class  = shift;
    my $caller = caller;

    strict->import;
    warnings->import;

    #base class
    {
        no strict 'refs';
        unshift @{"$caller\::ISA"}, $class;
    }

    #make action_table
    $caller->mk_classdata(action_table => []);
    #make attributes
    $caller->mk_classdata(attr => {});

    my @functions = qw(post get action run);
    #Export DSL
    for my $function ( @functions ) {
        $class->export_method($caller, $function);
    }
}

sub app {
    my $class = shift;
    return sub {
        my $env = shift;
        try {
            my $path = $env->{PATH_INFO};
            my $code;
            for my $action ( @{$class->action_table} ) {
                if ( $path =~ $action->{regex} && ( !defined $action->{method} || $action->{method} eq $env->{REQUEST_METHOD} ) ) {
                    $code = $action->{code};
                    last;
                }

            }
            Castella::Exception::HTTP::NotFound->throw unless $code;
            my $c = $class->new(
                req   => Castella::Request->new($env),
                res   => Castella::Response->new,
                stash => {},
                env   => $env,
            );
            $code->($c);
            my $templates = Data::Section::Simple->new($class)->get_data_section;
            my $xslate = Text::Xslate->new(
                path => $templates,
            );
            my $body = $xslate->render($path,$c->stash);
            Castella::Exception::HTTP::Success->throw(
                res => [
                    200,
                    ['Content-Type' => 'text/html'],
                    [$body]
                ],
            );
        } catch {
            my $e = shift;
            if( ref($e) && $e->isa('Castella::Exception::HTTP') ) {
                return $e->response;
            } else {
                #TODO Error handling
                Carp::confess($e);
            }

        }
    };
}

sub run (&) { $_[0] } ##no critic

sub add_action {
    my ($class, $regex, $options, $code, $method) = @_;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code , options => $options, method => $method};
}
sub action ($$) { ##no critic
    my ($regex, $code) = @_;
    my $class = caller;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code , template => $regex};
}

sub post ($$) { ##no critic
    my ($regex, $code) = @_;
    my $class = caller;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code, template => $regex, method => 'POST'};
}

sub get ($$) { ##no critic
    my ($regex, $code) = @_;
    my $class = caller;
    push @{$class->action_table}, {regex => qr{^$regex$} , code =>  $code, template => $regex, method => 'GET'};
}

1;
__END__
=encoding utf8

=head1 NAME

Castella - Sinatra like web application framework for PSGI/Plack

=head1 SYNOPSIS

  package  MyApp;
  use Castella;

  get '/' => run {
        my $self = shift;
        $self->stash->{message} = 'Hello Castella world!';
  };
  post '/' => run {
        my $self = shift;
        $self->stash->{message} = 'Castella is sweet!';
  };
  1;
  __DATA__
  @@ /
  <!DOCTYPE HTML>
  <html lang="ja">
  <head>
    <meta charset="UTF-8">
    <title>hogehoge</title>
  </head>
  <body>
    <: $message :>
    <form action="/" method="post">
        <button  type="submit">press me!</button>
    </form>
  </body>
  </html>

=head1 DESCRIPTION

Castella is web application framework which like Sinatra on Ruby

Write code easily!

=head1 METHODS

=head1 AUTHOR

Nishibayashi Takuji E<lt>takuji {at} senchan.jpE<gt>

=head1 SEE ALSO

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
